adserver/tasks.py

```diff
@@ -218,6 +232,128 @@ def daily_update_keywords(day=None):
+    """
+    start_date, end_date = _get_day(day)
+
+    log.info("Updating RegionTopic's for %s-%s", start_date, end_date)
+
+    # Remove all old impressions, because they are cumulative
+    RegionTopicImpression.objects.filter(
+        date__gte=start_date,
+        date__lt=end_date,
+    ).delete()
+
+    for impression_type in ["views"]:
+        queryset = _default_filters(impression_type, start_date, end_date)
+
+        unsold = {}
+        no_keywords = {}
+
+        for values in (
+            queryset.values("publisher", "advertisement", "keywords", "country")
+            .annotate(total=Count("country"))
+            .filter(total__gt=0)
+            .order_by("-total")
+            .values(
+                "publisher",
+                "advertisement",
+                "keywords",
+                "country",
+                "advertisement__flight__targeting_parameters",
+                "total",
+            )
+        ):
+            if not (
+                values["keywords"]
+                and values["country"]
+                and values["advertisement__flight__targeting_parameters"]
+            ):
+                continue
+
+            keywords = json.loads(values["keywords"])
+            country = values["country"]
+            publisher_keywords = set(keywords)
+
+            flight_targeting = json.loads(
+                values["advertisement__flight__targeting_parameters"]
+            )
+            flight_keywords = set(flight_targeting.get("include_keywords", {}))
+
+            matched_keywords = publisher_keywords & flight_keywords
+
+            topics = set()
+            for keyword in matched_keywords:
+                if keyword in data_science:
+                    topic = "data-science"
+                elif keyword in backend_web:
+                    topic = "backend-web"
+                elif keyword in frontend_web:
+                    topic = "frontend-web"
+                elif keyword in security_privacy:
+                    topic = "security-privacy"
+                elif keyword in devops:
+                    topic = "devops"
+                elif keyword in python:
+                    topic = "python"
+                elif keyword in blockchain:
+                    topic = "blockchain"
+                elif keyword in game_dev:
+                    topic = "game-dev"
+                else:
+                    topic = "other"
+                    if keyword in unsold:
+                        unsold[keyword] += 1
+                    else:
+                        unsold[keyword] = 1
+                topics.add(topic)
+
+            if not matched_keywords:
+                # No matched keywords
+                for keyword in publisher_keywords:
+                    if keyword in no_keywords:
+                        no_keywords[keyword] += 1
+                    else:
+                        no_keywords[keyword] = 1
+
+            if country in us_ca:
```

Comment:
Similar to topics, there will be overlap between regions. I think we need to have a regions set as well.

Reply:
I was assuming that region was exclusive. What is the case for overlapping regions?

Reply:
Our default targeting (NA/EU/Aus/NZ) includes Australia and New Zealand. However, they are also part of APAC. There are lots of semi-overlapping regions.

Reply:
Sure, but we don't have any overlapping regions in our data, so I'm not sure how I'd map something to multiple regions with the current data.

Reply:
I feel like we should probably keep this contract on the regions:

* Each country maps to exactly one subregion
* We can combine subregions into larger regions if we want (eg. Western Europe & Eastern Europe = Europe) -- but nothing would be tagged `Europe` in our data, just in aggregations.

Reply:
Ok.